src/System.Buffers.Primitives/System/Buffers/Buffer.cs

```diff
@@ -14,16 +14,14 @@ namespace System
     [DebuggerTypeProxy(typeof(BufferDebuggerView<>))]
     public struct Buffer<T>
     {
-        readonly OwnedBuffer<T> _owner;
-        readonly T[] _array;
+        readonly object _arrayOrOwnedBuffer;
```

Comment:
It would be good to add a comment that explains how this works, i.e. that _index high order bit is used to identify the modes.